src/utils/utils.cpp

```diff
@@ -59,16 +103,34 @@ bool openInNativeExtension(const QString &path) {
     //call ShellExecute internally
     return QDesktopServices::openUrl(QUrl::fromLocalFile(path));
 #elif defined(Q_WS_MAC)
-    QProcess open_process;
-    open_process.start(QLatin1String("open"), QStringList(path));
-    open_process.waitForFinished(-1);
-    return open_process.exitCode() == 0;
+    // mac's open program, it will fork to open the file in a subprocess
+    // so we will wait for it to check whether it succeeds or not
+    QProcess subprocess;
+    subprocess.start(QLatin1String("open"), QStringList(path));
+    subprocess.waitForFinished(-1);
+    return subprocess.exitCode() == 0;
 #elif defined(Q_WS_X11)
-    //xdg-open is sufficient
-    QProcess open_process;
-    open_process.start(QLatin1String("xdg-open"), QStringList(path));
-    open_process.waitForFinished(-1);
-    return open_process.exitCode() == 0;
+    // unlike mac's open program, xdg-open won't fork a new subprocess to open
+    // the file will block until the application returns, so we won't wait for it
+    // and we need another approach to check if it works
+
+    // find out if the file can be opened by xdg-open, xdg-mime
+    // usually they are installed in xdg-utils installed by default
+    QString mime_type;
+    if (!getMimeTypeFromXdgUtils(path, &mime_type))
+        return false;
+    // don't open this type of files from xdg-mime
+    if (mime_type == "application/octet-stream")
```

Comment:
I find out some interesting facts about `xdg-utils` from the comments of [chrome-wrapper](https://github.com/Chilledheart/chromium/blob/master/chrome/tools/build/linux/chrome-wrapper#L76) file from chromium project which has strong support for linux :

> We include some xdg utilities next to the binary, and we want to prefer them
> over the system versions when we know the system versions are very old.
 
From the results of searching `xdg-utils`-related topics, `xdg-mime` even might return with an empty output. And if it fails, it will fall back to `showInGraphicalShell`` .

Does this information help you?